content/5_Plat/Range_Sweep.mdx

```diff
@@ -74,111 +89,188 @@ int main() {
+		for (; ind > 1; ind /= 2) {
+			segtree[ind / 2] = std::min(segtree[ind], segtree[ind ^ 1]);
+		}
 	}
-	T query(int l, int r) {  // sum on interval [l, r]
-		T ra = ID, rb = ID;
-		for (l += n, r += n + 1; l < r; l /= 2, r /= 2) {
-			if (l & 1) ra = comb(ra, seg[l++]);
-			if (r & 1) rb = comb(seg[--r], rb);
+
+	T range_min(int start, int end) {
+		T min = DEFAULT;
+		for (start += len, end += len; start < end; start /= 2, end /= 2) {
+			if (start % 2 == 1) { min = std::min(min, segtree[start++]); }
+			if (end % 2 == 1) { min = std::min(min, segtree[--end]); }
 		}
-		return comb(ra, rb);
+		return min;
 	}
 };
+// EndCodeSnip
+
+struct Point {
+	int x, y;
+	int i;
+	bool is_start;
 
-Seg<int> S;
-int N, P;
-int ans[MX];
-vi distinct_y;
+	Point(int x, int y, int i, bool is_start) : x(x), y(y), i(i), is_start(is_start) {}
 
-int query_min(int ind) { return S.query(0, ind); }
-void ins(int ind, int val) { S.upd(ind, val); }
-int y_index(int y) { return lb(all(distinct_y), y) - begin(distinct_y); }
+	bool operator<(const Point &p) {
+		if (x == p.x) { return y < p.y; }
+		return x < p.x;
+	}
+};
+
+vector<int> distinct_y;
+
+int y_index(int y) {
+	return lower_bound(begin(distinct_y), end(distinct_y), y) -
+	       begin(distinct_y);
+}
 
 int main() {
-	setIO("boards");
-	cin >> N >> P;
-	vector<pair<pair<int, int>, pair<int, int>>> ev;
-	for (int i = 0; i < P; ++i) {
-		pair<int, int> a, b;
-		cin >> a.f >> a.s >> b.f >> b.s;
-		ev.push_back({a, {i, -1}});  // start point
-		ev.push_back({b, {i, 1}});   // end point
-		distinct_y.pb(a.s);
-		distinct_y.pb(b.s);
+	freopen("boards.in", "r", stdin);
+	int n, p;
+	cin >> n >> p;
+
+	vector<Point> events;
+	for (int i = 0; i < p; ++i) {
+		int x1, y1, x2, y2;
+		cin >> x1 >> y1 >> x2 >> y2;
+		events.emplace_back(x1, y1, i, true);  // start point
+		events.emplace_back(x2, y2, i, false);   // end point
+		distinct_y.push_back(y1);
+		distinct_y.push_back(y2);
 	}
-	sort(all(distinct_y));
-	sort(begin(ev), end(ev));
-	S.init(2 * P);
-	ins(0, 0);
-	for (auto &t : ev) {
-		if (t.s.s == -1) {
-			ans[t.s.f] = t.f.f + t.f.s + query_min(y_index(t.f.s));
+
+	sort(begin(distinct_y), end(distinct_y));
+	sort(begin(events), end(events));
+
+	vector<int> ans(p);
+	MinSegmentTree<int> segtree(2 * p);
+	segtree.set(0, 0);
+	for (const Point& p : events) {
+		if (p.is_start) {
+			ans[p.i] = p.x + p.y + segtree.range_min(0, y_index(p.y) + 1);
 		} else {
-			ins(y_index(t.f.s), ans[t.s.f] - t.f.f - t.f.s);
+			segtree.set(y_index(p.y), ans[p.i] - p.x - p.y);
 		}
 	}
-	cout << query_min(2 * P - 1) + 2 * N;
+
+	freopen("boards.out", "w", stdout);
+	// adds 2N to include the transition
+	// from the last springboard to the endpoint
+	cout << segtree.range_min(0, 2 * p) + 2 * n;
 }
 ```
 
+
+</CPPSection>
+</LanguageSection>
+
+Bonus: Rather than thinking of this problem as minimizing the distance walked, think of this problem as maximizing the distance covered by springboards to simplify the DP transition.
```

Comment:
so is this a sorta "exercise left to the reader"' kinda thing

Reply:
yes

Reply:
@ryanchou-dev thoughts? i was never a fan of these exercises without matching solutions but you may think otherwise

Reply:
i think its fine

i was never a fan of spoonfeeding every single detail to the reader

Reply:
well do you have anywhere they can go if they can't figure this out?

Reply:
this bonus is not much different from the given sol
imo doing it in this way is more educational so the reader can see for themselves how a subtle change in perspective can change a sol
the idea isnt complicated

Reply:
they can just skip it lol
i wouldnt add a solution just to help a small part of the population if it decreases the educational value for everyone else
all of the hard ideas are given in the given sol

Reply:
how...does adding a solution decrease the educational value?

Reply:
cp is about problem solving.....
cses problems dont have editorials for a reason
in this case there are even less reasons why an editoral is needed if all the core ideas are already given
https://usaco.guide/plat/convex-hull-trick
look at the blue alternative solution box

Reply:
> cses problems dont have editorials for a reason

in what world is this a good thing
look just either write the solution or just don't include it at all

Reply:
you would be arguing against virtually every CP website including this one

Reply:
im not going to argue anymore